Duplication and atlas update

Hakazaba opened the pull request:
Duplication now will be taken account of during the generation of the atlas. This may need an update of Inochi session to be compatible and needs testing.

There are some variables that may not be currently copied over. Someone more familiar with the functions of composites for instance, should look over the variables to make sure everything important is copied.

Hakazaba committed:
Added Duplicate function

Added to the nodes right click menu. There may be some variables that were missed, particularly animations.

Hakazaba committed:
Update nodes.d

Hakazaba committed:
Update nodes.d

Hakazaba committed:
Atlas update

Atlassing switched to per texture instead of based on UUid

Hakazaba committed:
User friendliness

Updated the duplication to place the new object next in the order of children

LunaTheFoxgirl commented on source/creator/actions/node.d:
```diff
@@ -332,6 +332,76 @@ void incAddChildWithHistory(Node n, Node to, string name=null) {
+Node recursiveDuplicate(Node n){
+    Node x;
+
+    if ((cast(Part) n) is null) {
```
Please follow the code-style used throughout the rest of Inochi2D, this is difficult to read.

Hakazaba replied:
I'll try to clean it up a bit tomorrow

LunaTheFoxgirl commented on source/creator/actions/node.d:
```diff
@@ -332,6 +332,76 @@ void incAddChildWithHistory(Node n, Node to, string name=null) {
+
+ 
+    foreach (child; n.children())
+    {
```
This curlybrace should be on the same line as the foreach.

LunaTheFoxgirl commented on source/creator/panels/nodes.d:
```diff
@@ -165,7 +167,25 @@ protected:
                 if (igMenuItem(__("MeshGroup"), "", false, true)) incAddChildWithHistory(new MeshGroup(cast(Node)null), n);
-
+                
+                if((cast(Camera)n is null) & (cast(Puppet)n is null)){
```
This could be simplified to
```d
if (cast(Camera)n || cast(Puppet)n)
```

Hakazaba replied:
We only want to proceed if they're both null. If they aren't then we might end up trying to copy the puppet or camera.

LunaTheFoxgirl replied:
In which case just do !cast; that will be a is-null check

LunaTheFoxgirl reviewed (changes requested):
Good work, however there's some cleanup to do before this can be merged.

Hakazaba committed:
Cleaned the code

Sorry about that, To be totally honest, I wrote that before I even discovered not was !is. !cast was a game changer
